components/engine_rocks/src/event_listener.rs

```diff
@@ -190,7 +190,7 @@ impl RocksPersistenceListener {
 impl rocksdb::EventListener for RocksPersistenceListener {
     fn on_memtable_sealed(&self, info: &MemTableInfo) {
         self.0
-            .on_memtable_sealed(info.cf_name().to_string(), info.earliest_seqno());
+            .on_memtable_sealed(info.cf_name().to_string(), info.first_seqno());
```

Comment:
`earliest_seqno` is used on purpose as I have found `first_seqno` not always be the smallest seqno.

Reply:
You have to be more specific, e.g. reproduce it in unit test. And then we'll see if that particular situation can be fixed. Because RocksDB code doesn't guarantee `earliest_seqno_` to be smaller than `first_seqno_`, in fact in several places it attempts to make them equal. It's dangerous to assume it.

Reply:
Just add some logs in `on_memtable_sealed` in event_listener.rs:192 and run test case `test_persistence_listener` should show you the difference.

> Because RocksDB code doesn't guarantee earliest_seqno_ to be smaller than first_seqno_

But the API docs say it's guaranteed to be smallest.

> // Returns the sequence number that is guaranteed to be smaller than or equal
  // to the sequence number of any key that could be inserted into this
  // memtable. It can then be assumed that any write with a larger(or equal)
  // sequence number will be present in this memtable or a later memtable.
  //
  // If the earliest sequence number could not be determined,
  // kMaxSequenceNumber will be returned.

Reply:
I know they are different, that's why I changed [components/engine_traits/src/flush.rs](https://github.com/tikv/tikv/pull/14253/files#diff-39f9233251cbe78f09636b94265347b0739864b2222ca76dbc86d8c66f3735f3).

> But the API docs say it's guaranteed to be smallest.

It's guaranteed to be <= the smallest seqno in memtable. But in our code, we assume "flushed largest_seqno equals to earliest_seqno of next memtable.".

Reply:
Note we don't assume "flushed largest_seqno equals to earliest_seqno of next memtable", but we assume "flushed largest_seqno is not larger than earliest_seqno of next memtable".